### PR TITLE
Single code path for value transformations

### LINKED ISSUE #728: Internal refactor: Single codepath for transformations
Part of #713

(Claus - 2) Make a single code path for value transformations going in and out of C-interop layer. We currently have multiple places where C-API values are translated between C-interop types and user facing types. Most notably the various differentiated methods in RealmObjectHelper and ElementConverters in RealmListInternal.kt, but the need will increase with support for sets, mixed, and dictionaries.

- Required as part of a public Type converter API.
- Do it after adding composite internal object (#727)

### PR DESCRIPTION
**Pull request by rorbech**
This PR isolates implementations of conversion between public types and cinterop types into `Converters.kt`.

Implementation now uses a
- _public type_ in the user facing API (allowing _public_ `Byte`, `Char`, `Short` and `Int` types for `RealmStorageType.INT`)
- _storage types_  as a Kotlin instance corresponding to the Core's property type. Mapping is documented in https://github.com/realm/realm-kotlin/blob/master/packages/library-base/src/commonMain/kotlin/io/realm/schema/RealmStorageType.kt
- `cinterop` types which are Kotlin implementations of `cinterop`'s `realm_value_t`s. Currently it is only [`RealmObjectInterop`](https://github.com/realm/realm-kotlin/blob/master/packages/cinterop/src/commonMain/kotlin/io/realm/internal/interop/RealmObjectInterop.kt) and [`Timestamp`](https://github.com/realm/realm-kotlin/blob/master/packages/cinterop/src/commonMain/kotlin/io/realm/internal/interop/Timestamp.kt) that has special implementations. The rest are just passed as their raw _storage type_, i.e. `String`, `Long`, etc.  

The `cinterop`-types are passed in and out of cinterop as a `value class RealmValue`, which basically wraps the `cinterop` type as an Any/Mixed-type (😉 😅)

Converters for the various public-to-cinterop conversion are in [`Converters.kt`](https://github.com/realm/realm-kotlin/blob/cr/common-realm-value-conversion/packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt) and is split into a two step process of
- Converting between a _public type_ and a _storage type_
- Converting between a _storage type_ and the `cinterop` types

to allow future reuse of the infrastructure for custom type maps where the public-to-storage type conversion can be provided by the user. 

All current conversion functions are currently implemented as top level functions in `Converters.kt` to allow inlining in the accessorts injected by the compiler plugin along with inlining them in the individual `RealmValueConverter`s that is used throughout the `RealmList`/`RealmQuery`-implementations to fix a `RealmList`/`RealmQuery` with a specific converter upon instantiation. 

Considerations:
TODO/Futures:
- [x] Fix native tests 😅 
- [ ] **Future** Consider reimplementing `RealmValue` as a wrapper on top of `realm_value_t` to allow returning those from `cinterop` without converting values between public types and `realm_value_t` unless really needed to allow optimization when passing `Any/Mixed` from on internal API to another, etc. 
- [ ] **Future?** `RealmObject`s are currently returned as `Link`s when returned from the C-API as a `RealmValue`, while it is passed to the C-API as a `RealmObjectInterop`. We could streamline this, but didn't want to rework it as part of this PR. We could also make the actual native pointer lazy-loaded from the link, thus only resolving the actual native reference until actually needed similarly to what is proposed for `RealmValue/realm_value_t` below

Closes #728

### CODE REVIEW

**Review by cmelchior — COMMENTED** [review]
This PR could probably use a walkthrough in person :D 

Generally, this looks good I think, but I have some architectural concerns about the number of converter types and how they are being used.

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+import io.realm.dynamic.DynamicRealmObject
+import io.realm.internal.interop.Link
+import io.realm.internal.interop.RealmValue
+import io.realm.internal.interop.Timestamp
+import io.realm.internal.platform.realmObjectCompanionOrNull
+import kotlin.native.concurrent.SharedImmutable
+import kotlin.reflect.KClass
+
+// Interface for converting storage types (Kotlin representation of Core values) to C-API RealmValue
+public interface StorageTypeConverter<T> {
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt]: Not any more 👍

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    public fun toPublic(value: S?): T?
+}
+
+// Interface for converting public type to C-API RealmValue
+public interface RealmValueConverter<T> {
+    public fun publicToRealmValue(value: T?): RealmValue
+    public fun realmValueToPublic(realmValue: RealmValue): T?
+}
+
+public interface ConverterInternal<T, S> :
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt]: Renamed it to `CompositeConverter` and added some brief comments.

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    RealmValueConverter<T>, PublicConverter<T, S>, StorageTypeConverter<S> {
+    override fun publicToRealmValue(value: T?): RealmValue = toRealmValue(fromPublic(value))
+    override fun realmValueToPublic(realmValue: RealmValue): T? =
+        toPublic(fromRealmValue(realmValue))
+}
+
+// Converter with default identity conversion implementation
+internal interface IdentityConverter<T> : ConverterInternal<T, T>, StorageTypeConverter<T> {
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt]: I like `PassThrough` ... updated it 👍

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    mapOf<KClass<*>, ConverterInternal<*, *>>(
+        Byte::class to ByteConverter,
+        Char::class to CharConverter,
+        Short::class to ShortConverter,
+        Int::class to IntConverter,
+        RealmInstant::class to RealmInstantConverter
+    ).withDefault { StaticIdentityConverter }
+
+// Dynamic default primitive value converter to translate primary keys and query arguments to RealmValues
+public object RealmValueArgumentConverter : RealmValueConverter<Any?> {
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt]: No, it ended up not needing to implement the interface. Also added a convenience method to convert a list of arguments.

**Comment by rorbech** [comment]
> This PR could probably use a walkthrough in person :D
> 
> Generally, this looks good I think, but I have some architectural concerns about the number of converter types and how they are being used.

I address all the specific comments ... let's take the remaining architectural concerns in the meeting tomorrow.

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt [hunk 5]
```diff
@@ -239,15 +199,15 @@ internal object RealmObjectHelper {
...
-            DynamicRealmObject::class -> getObjectByKey<DynamicRealmObject>(obj, propertyInfo.key)
-            DynamicMutableRealmObject::class -> getObjectByKey<DynamicMutableRealmObject>(
-                obj,
-                propertyInfo.key
-            )
-            else -> getValueByKey<R>(obj, propertyInfo.key)
+            DynamicRealmObject::class,
+            DynamicMutableRealmObject::class ->
+                realmValueToRealmObject(realmValue, clazz as KClass<out RealmObject>, obj.mediator, obj.owner)
+            else -> primitiveTypeConverters.getValue(clazz).realmValueToPublic(realmValue)
```
Comment (cmelchior) [inline packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt:210]: Do we have a test for what happens if no converter is found here?

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+public inline fun anyToRealmValue(value: Any?): RealmValue = RealmValue(value)
+
+// Interface for converting public types to storage types (Kotlin representation of Core values)
+public interface PublicConverter<T, S> {
+    public fun fromPublic(value: T?): S?
+    public fun toPublic(value: S?): T?
+}
+
+// Interface for converting between public type and C-API input/output-values (RealmValue)
+internal interface RealmValueConverter<T> {
```
Comment (cmelchior) [inline packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt:47]: Maybe add a comment why we want to have this alongside the other two convert types?

**Review by cmelchior — APPROVED** [review]
Really nice improvement 💯  . I only had a few minor comments and improvement suggestions, but otherwise it looks good 👍.

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt (hunk 5 above)
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt:210]: `getValue()` returns the `StaticPassThroughConverter` as supplied with `withDefault` when initializing `primitiveTypeConverters`. This means that tests for all the types without explicit converts in `primitiveTypeConverts` are already exercising this.

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+    override inline fun fromRealmValue(realmValue: RealmValue): RealmInstant? =
+        realmValueToRealmInstant(realmValue)
+}
+
+// Top level method to allow inlining from compiler plugin
+public inline fun realmValueToRealmInstant(realmValue: RealmValue): RealmInstant? =
+    realmValue.value?.let { RealmInstantImpl(it as Timestamp) }
+
+@SharedImmutable
+internal val primitiveTypeConverters: Map<KClass<*>, ConverterInternal<*, *>> =
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt]: There is already a default ... from `withDefault(StaticPassThroughConverter)`.

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+public inline fun anyToRealmValue(value: Any?): RealmValue = RealmValue(value)
+
+// Interface for converting public types to storage types (Kotlin representation of Core values)
+public interface PublicConverter<T, S> {
+    public fun fromPublic(value: T?): S?
+    public fun toPublic(value: S?): T?
+}
+
+// Interface for converting public type to C-API RealmValue
+public interface RealmValueConverter<T> {
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt]: I have added some more comments around it.

File: packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt (hunk 6 above)
Comment (rorbech) [reply packages/library-base/src/commonMain/kotlin/io/realm/internal/Converters.kt]: 👍 I tried to give an overall brief of the converter-concept/motivation at the top of the file and relate the individual interfaces to that. Hopefully it should be more clear now.

**Review by cmelchior — APPROVED** [review]

**Pull request merged**